GDJS/tests/tests/runtimescene.js

```diff
@@ -83,100 +83,155 @@ describe('gdjs.RuntimeScene integration tests', function () {
+      expect(object2.getLayer()).to.be('');
+      expect(object3.getLayer()).to.be('MyOtherLayer');
+      expect(runtimeScene.hasLayer('')).to.be(true);
+      expect(runtimeScene.hasLayer('MyLayer')).to.be(false);
+      expect(runtimeScene.hasLayer('MyOtherLayer')).to.be(true);
     });
 
-    expect(runtimeScene.hasLayer('')).to.be(true);
-    expect(runtimeScene.hasLayer('MyLayer')).to.be(true);
-    expect(runtimeScene.hasLayer('MyOtherLayer')).to.be(false);
-
-    const object1 = runtimeScene.createObject('MyObject');
-    const object2 = runtimeScene.createObject('MyObject');
-    const object3 = runtimeScene.createObject('MyObject');
-    if (!object1 || !object2 || !object3) {
-      throw new Error('object should have been created');
-    }
-    object2.setLayer('MyLayer');
-
-    runtimeScene.addLayer({
-      name: 'MyOtherLayer',
-      visibility: true,
-      cameras: [],
-      effects: [],
-      isLightingLayer: false,
-      followBaseLayerCamera: false,
-      ambientLightColorR: 128,
-      ambientLightColorG: 128,
-      ambientLightColorB: 128,
+    it('should compute layers highest z orders correctly', () => {
+      expect(runtimeScene.hasLayer('')).to.be(true);
+      expect(runtimeScene.hasLayer('MyLayer')).to.be(true);
+
+      const object1 = runtimeScene.createObject('MyObject');
+      const object2 = runtimeScene.createObject('MyObject');
+      const object3 = runtimeScene.createObject('MyObject');
+      if (!object1 || !object2 || !object3) {
+        throw new Error('object should have been created');
+      }
+      expect(object2.getZOrder()).to.be(0);
+      object2.setLayer('MyLayer');
+
+      // Layers highest Z orders should stay at 0 if objects did not change Z their order
+      runtimeScene._updateObjectsPreRender();
+      expect(runtimeScene.getLayer('MyLayer').getHighestZOrder()).to.be(0);
+      expect(runtimeScene.getLayer('').getHighestZOrder()).to.be(0);
+
+      object2.setZOrder(8);
+
+      // Check highest Z order has been correctly updated for layers and they are stable
+      runtimeScene._updateObjectsPreRender();
+      expect(runtimeScene.getLayer('MyLayer').getHighestZOrder()).to.be(8);
+      expect(runtimeScene.getLayer('').getHighestZOrder()).to.be(0);
+      runtimeScene._updateObjectsPreRender();
+      expect(runtimeScene.getLayer('MyLayer').getHighestZOrder()).to.be(8);
+      expect(runtimeScene.getLayer('').getHighestZOrder()).to.be(0);
+
+      // Change Z orders for default layer
+      object1.setZOrder(13);
+      object3.setZOrder(25);
+
+      runtimeScene._updateObjectsPreRender();
+      expect(runtimeScene.getLayer('MyLayer').getHighestZOrder()).to.be(8);
+      expect(runtimeScene.getLayer('').getHighestZOrder()).to.be(25);
+
+      // Check highest z orders come back to 0 after objects deletion
+      runtimeScene.markObjectForDeletion(object1);
+      runtimeScene.markObjectForDeletion(object2);
+      runtimeScene.markObjectForDeletion(object3);
```

Comment:
It could be interesting to only remove the object with Z order = 25 first.

What about moving one instance in front of the top one?